File: src/main/java/org/ldk/structs/FundingInfo.java
package org.ldk.structs;

import org.ldk.impl.bindings;
import org.ldk.enums.*;
import org.ldk.util.*;
import java.util.Arrays;
import java.lang.ref.Reference;
import javax.annotation.Nullable;


/**
 * `FundingInfo` holds information about a channel's funding transaction.
 * 
 * When LDK is set to manual propagation of the funding transaction
 * (via [`ChannelManager::unsafe_manual_funding_transaction_generated`),
 * LDK does not have the full transaction data. Instead, the `OutPoint`
 * for the funding is provided here.
 * 
 * [`ChannelManager::unsafe_manual_funding_transaction_generated`]: crate::ln::channelmanager::ChannelManager::unsafe_manual_funding_transaction_generated
 */
@SuppressWarnings("unchecked") // We correctly assign various generic arrays
public class FundingInfo extends CommonBase {
	private FundingInfo(Object _dummy, long ptr) { super(ptr); }
	@Override @SuppressWarnings("deprecation")
	protected void finalize() throws Throwable {
		super.finalize();
		if (ptr != 0) { bindings.FundingInfo_free(ptr); }
	}
	static FundingInfo constr_from_ptr(long ptr) {
		bindings.LDKFundingInfo raw_val = bindings.LDKFundingInfo_ref_from_ptr(ptr);
		if (raw_val.getClass() == bindings.LDKFundingInfo.Tx.class) {
			return new Tx(ptr, (bindings.LDKFundingInfo.Tx)raw_val);
		}
		if (raw_val.getClass() == bindings.LDKFundingInfo.OutPoint.class) {
			return new OutPoint(ptr, (bindings.LDKFundingInfo.OutPoint)raw_val);
		}
		assert false; return null; // Unreachable without extending the (internal) bindings interface
	}

	/**
	 * The full funding `Transaction`.
	 */
	public final static class Tx extends FundingInfo {
		/**
		 * The funding transaction
		*/
		public final byte[] transaction;
		private Tx(long ptr, bindings.LDKFundingInfo.Tx obj) {
			super(null, ptr);
			this.transaction = obj.transaction;
		}
	}
	/**
	 * The `OutPoint` of the funding.
	 */
	public final static class OutPoint extends FundingInfo {
		/**
		 * The outpoint of the funding
		*/
		public final org.ldk.structs.OutPoint outpoint;
		private OutPoint(long ptr, bindings.LDKFundingInfo.OutPoint obj) {
			super(null, ptr);
			long outpoint = obj.outpoint;
			org.ldk.structs.OutPoint outpoint_hu_conv = null; if (outpoint < 0 || outpoint > 4096) { outpoint_hu_conv = new org.ldk.structs.OutPoint(null, outpoint); }
			if (outpoint_hu_conv != null) { outpoint_hu_conv.ptrs_to.add(this); };
			this.outpoint = outpoint_hu_conv;
		}
	}
	long clone_ptr() {
		long ret = bindings.FundingInfo_clone_ptr(this.ptr);
		Reference.reachabilityFence(this);
		return ret;
	}

	/**
	 * Creates a copy of the FundingInfo
	 */
	public FundingInfo clone() {
		long ret = bindings.FundingInfo_clone(this.ptr);
		Reference.reachabilityFence(this);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.FundingInfo ret_hu_conv = org.ldk.structs.FundingInfo.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(this); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new Tx-variant FundingInfo
	 */
	public static FundingInfo tx(byte[] transaction) {
		long ret = bindings.FundingInfo_tx(transaction);
		Reference.reachabilityFence(transaction);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.FundingInfo ret_hu_conv = org.ldk.structs.FundingInfo.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new OutPoint-variant FundingInfo
	 */
	public static FundingInfo out_point(org.ldk.structs.OutPoint outpoint) {
		long ret = bindings.FundingInfo_out_point(outpoint.ptr);
		Reference.reachabilityFence(outpoint);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.FundingInfo ret_hu_conv = org.ldk.structs.FundingInfo.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.add(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Checks if two FundingInfos contain equal inner contents.
	 * This ignores pointers and is_owned flags and looks at the values in fields.
	 */
	public boolean eq(org.ldk.structs.FundingInfo b) {
		boolean ret = bindings.FundingInfo_eq(this.ptr, b.ptr);
		Reference.reachabilityFence(this);
		Reference.reachabilityFence(b);
		return ret;
	}

	@Override public boolean equals(Object o) {
		if (!(o instanceof FundingInfo)) return false;
		return this.eq((FundingInfo)o);
	}
	/**
	 * Serialize the FundingInfo object into a byte array which can be read by FundingInfo_read
	 */
	public byte[] write() {
		byte[] ret = bindings.FundingInfo_write(this.ptr);
		Reference.reachabilityFence(this);
		return ret;
	}

	/**
	 * Read a FundingInfo from a byte array, created by FundingInfo_write
	 */
	public static Result_FundingInfoDecodeErrorZ read(byte[] ser) {
		long ret = bindings.FundingInfo_read(ser);
		Reference.reachabilityFence(ser);
		if (ret >= 0 && ret <= 4096) { return null; }
		Result_FundingInfoDecodeErrorZ ret_hu_conv = Result_FundingInfoDecodeErrorZ.constr_from_ptr(ret);
		return ret_hu_conv;
	}

}

File: c_sharp/src/org/ldk/structs/FundingInfo.cs
using org.ldk.impl;
using org.ldk.enums;
using org.ldk.util;
using System;

namespace org { namespace ldk { namespace structs {

/**
 * `FundingInfo` holds information about a channel's funding transaction.
 * 
 * When LDK is set to manual propagation of the funding transaction
 * (via [`ChannelManager::unsafe_manual_funding_transaction_generated`),
 * LDK does not have the full transaction data. Instead, the `OutPoint`
 * for the funding is provided here.
 * 
 * [`ChannelManager::unsafe_manual_funding_transaction_generated`]: crate::ln::channelmanager::ChannelManager::unsafe_manual_funding_transaction_generated
 */
public class FundingInfo : CommonBase {
	protected FundingInfo(object _dummy, long ptr) : base(ptr) { }
	~FundingInfo() {
		if (ptr != 0) { bindings.FundingInfo_free(ptr); }
	}

	internal static FundingInfo constr_from_ptr(long ptr) {
		long raw_ty = bindings.LDKFundingInfo_ty_from_ptr(ptr);
		switch (raw_ty) {
			case 0: return new FundingInfo_Tx(ptr);
			case 1: return new FundingInfo_OutPoint(ptr);
			default:
				throw new ArgumentException("Impossible enum variant");
		}
	}

	/** A FundingInfo of type Tx */
	public class FundingInfo_Tx : FundingInfo {
		/**
		 * The funding transaction
		 */
		public byte[] transaction;
		internal FundingInfo_Tx(long ptr) : base(null, ptr) {
			long transaction = bindings.LDKFundingInfo_Tx_get_transaction(ptr);
			byte[] transaction_conv = InternalUtils.decodeUint8Array(transaction);
			this.transaction = transaction_conv;
		}
	}
	/** A FundingInfo of type OutPoint */
	public class FundingInfo_OutPoint : FundingInfo {
		/**
		 * The outpoint of the funding
		 */
		public org.ldk.structs.OutPoint outpoint;
		internal FundingInfo_OutPoint(long ptr) : base(null, ptr) {
			long outpoint = bindings.LDKFundingInfo_OutPoint_get_outpoint(ptr);
			org.ldk.structs.OutPoint outpoint_hu_conv = null; if (outpoint < 0 || outpoint > 4096) { outpoint_hu_conv = new org.ldk.structs.OutPoint(null, outpoint); }
			if (outpoint_hu_conv != null) { outpoint_hu_conv.ptrs_to.AddLast(this); };
			this.outpoint = outpoint_hu_conv;
		}
	}
	internal long clone_ptr() {
		long ret = bindings.FundingInfo_clone_ptr(this.ptr);
		GC.KeepAlive(this);
		return ret;
	}

	/**
	 * Creates a copy of the FundingInfo
	 */
	public org.ldk.structs.FundingInfo clone() {
		long ret = bindings.FundingInfo_clone(this.ptr);
		GC.KeepAlive(this);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.FundingInfo ret_hu_conv = org.ldk.structs.FundingInfo.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(this); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new Tx-variant FundingInfo
	 */
	public static org.ldk.structs.FundingInfo tx(byte[] transaction) {
		long ret = bindings.FundingInfo_tx(InternalUtils.encodeUint8Array(transaction));
		GC.KeepAlive(transaction);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.FundingInfo ret_hu_conv = org.ldk.structs.FundingInfo.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Utility method to constructs a new OutPoint-variant FundingInfo
	 */
	public static org.ldk.structs.FundingInfo out_point(org.ldk.structs.OutPoint outpoint) {
		long ret = bindings.FundingInfo_out_point(outpoint.ptr);
		GC.KeepAlive(outpoint);
		if (ret >= 0 && ret <= 4096) { return null; }
		org.ldk.structs.FundingInfo ret_hu_conv = org.ldk.structs.FundingInfo.constr_from_ptr(ret);
		if (ret_hu_conv != null) { ret_hu_conv.ptrs_to.AddLast(ret_hu_conv); };
		return ret_hu_conv;
	}

	/**
	 * Checks if two FundingInfos contain equal inner contents.
	 * This ignores pointers and is_owned flags and looks at the values in fields.
	 */
	public bool eq(org.ldk.structs.FundingInfo b) {
		bool ret = bindings.FundingInfo_eq(this.ptr, b.ptr);
		GC.KeepAlive(this);
		GC.KeepAlive(b);
		return ret;
	}

	public override bool Equals(object o) {
		if (!(o is FundingInfo)) return false;
		return this.eq((FundingInfo)o);
	}
	/**
	 * Serialize the FundingInfo object into a byte array which can be read by FundingInfo_read
	 */
	public byte[] write() {
		long ret = bindings.FundingInfo_write(this.ptr);
		GC.KeepAlive(this);
		if (ret >= 0 && ret <= 4096) { return null; }
		byte[] ret_conv = InternalUtils.decodeUint8Array(ret);
		return ret_conv;
	}

	/**
	 * Read a FundingInfo from a byte array, created by FundingInfo_write
	 */
	public static org.ldk.structs.Result_FundingInfoDecodeErrorZ read(byte[] ser) {
		long ret = bindings.FundingInfo_read(InternalUtils.encodeUint8Array(ser));
		GC.KeepAlive(ser);
		if (ret >= 0 && ret <= 4096) { return null; }
		Result_FundingInfoDecodeErrorZ ret_hu_conv = Result_FundingInfoDecodeErrorZ.constr_from_ptr(ret);
		return ret_hu_conv;
	}

}
} } }
